// @wix/flynt-rules/dist/catalog/deprecations/invalid-deprecate-target-removal-date/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
const array_utils_1 = require("../../array-utils");
function assertTargetRemovalDate2(ctx, type, targetRemovalDateOptions) {
    if (targetRemovalDateOptions) {
        const options = (0, array_utils_1.isArray)(targetRemovalDateOptions) ? targetRemovalDateOptions : [targetRemovalDateOptions];
        options.forEach(opt => {
            if (opt.value.targetRemovalDate) {
                ctx.assert((0, fqdn_utils_1.isValidDate)(opt.value.targetRemovalDate), {
                    errorAt: opt,
                    message: `the value of the 'target_removal_date' in the deprecate option of the ${type.typeName.toString().toLowerCase()} '${(0, fqdn_utils_1.toLocalFqn)(type.fqn)}' has an invalid date format '${opt.value.targetRemovalDate}'`
                });
            }
        });
    }
}
exports.rule = {
    description: 'A deprecated target_removal_date must be in a valid date format',
    severity: flynt_engine_1.Severity.Error,
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/protobuf#platformization-guidelines_protobuf_deprecation',
    run(ctx) {
        ctx.visit({
            visitService(service) {
                assertTargetRemovalDate2(ctx, service, service.options.getTypedOption(wix_protos_1.wix.api.service_deprecated));
                assertTargetRemovalDate2(ctx, service, service.options.getTypedOption(wix_protos_1.wix.api.event_deprecated));
            },
            visitField(field) {
                assertTargetRemovalDate2(ctx, field, field.options.getTypedOption(wix_protos_1.wix.api.field_deprecated));
            },
            visitMessage(message) {
                assertTargetRemovalDate2(ctx, message, message.options.getTypedOption(wix_protos_1.wix.api.message_deprecated));
            },
            visitMethod(method) {
                assertTargetRemovalDate2(ctx, method, method.options.getTypedOption(wix_protos_1.wix.api.deprecated));
            }
        });
    }
};
