// @wix/flynt-rules/dist/catalog/field-validations/missing-array-limits/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'A repeated field must have a maxSize limit',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/design-to-exposure-process#platformization-guidelines_design-to-exposure-process_full-process',
    breakIfNew: true,
    run(ctx) {
        const responses = new Set();
        ctx.visit({
            visitMethod(method) {
                responses.add(method.response);
            }
        });
        ctx.visit({
            visitMessage(message) {
                if (responses.has(message)) {
                    return false;
                }
            },
            visitField(field) {
                const maxSizeValidation = field.options.getTypedOption(wix_protos_1.wix.api.maxSize);
                if (field.fieldType.kind === flynt_engine_1.FieldKind.Repeated) {
                    ctx.assert(maxSizeValidation, {
                        errorAt: field,
                    });
                }
            }
        });
    }
};
