// @wix/flynt-rules/dist/catalog/spi/new-method-missing-feature-toggle/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'a new SPI method must have `wix.spi.feature_toggle` option',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface#platformization-guidelines_spi---service-provider-interface_spi---service-provider-interface',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visitAdded({
            shouldVisitService(service) {
                return (0, fqdn_utils_1.isSpiService)(service);
            },
            visitMethod(method) {
                const featureToggles = method.options.getTypedOption(wix_protos_1.wix.spi.feature_toggle);
                ctx.assert(featureToggles && featureToggles.value, {
                    errorAt: method,
                });
            }
        });
    }
};
