// @wix/flynt-rules/dist/catalog/field-validations/new-field-last-in-order/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
exports.rule = {
    description: 'A new field must be appended at the end of message field list',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/api-design-patterns#platformization-guidelines_api-design-patterns_resilience-to-added-fields',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if (!(0, fqdn_utils_1.isServiceRelevantForAutoGeneratedClient)(service)) {
                    return false;
                }
            },
            visitMethod(method) {
                if ((0, fqdn_utils_1.isMethodRelevantForAutoGeneratedClient)(method)) {
                    ctx.visitAddedFrom(method.request, {
                        visitField(field) {
                            const index = field.parent.fields.indexOf(field);
                            const prevMessage = ctx.prevSt.getMessage(field.parent.fqn);
                            ctx.assert(index >= prevMessage.fields.length, {
                                errorAt: field,
                                message: `the new field '${field.name}' must be appended at the end of '${prevMessage.name}' request message`
                            });
                        }
                    });
                }
            }
        });
    }
};
