// @wix/flynt-rules/dist/catalog/events/invalid-segment/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
const missing_segment_1 = require("../missing-segment");
const callback_1 = require("@wix/wix-protos/typed/wix/api/callback");
const { topicMappings } = require('@wix/greyhound-legacy-topics');
function getExpectedSegment(fqdn) {
    const topic = (0, missing_segment_1.domainEventTopic)(fqdn);
    return segmentNameToEnum[topicMappings[topic]];
}
const segmentNameToEnum = {
    'Users1': callback_1.Segment.Segment.USERS,
    'Public1': callback_1.Segment.Segment.PUBLIC,
    'Others1': callback_1.Segment.Segment.OTHERS,
};
exports.rule = {
    description: 'Invalid segment was used for a topic which was already migrated to a specific segment',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitMessage(message) {
                const entityOption = message.options.getTypedOption(wix_protos_1.wix.api.entity);
                if (entityOption) {
                    if (entityOption.value.segment) {
                        const fqdn = entityOption.value.oldFqdnForBackwardsCompatibility ?? entityOption.value.fqdn;
                        const expectedSegment = getExpectedSegment(fqdn);
                        if (expectedSegment) {
                            ctx.assert(expectedSegment === entityOption.value.segment, {
                                errorAt: entityOption,
                                message: `'${fqdn}' is only allowed to use the segment ${callback_1.Segment.Segment[expectedSegment]}`
                            });
                        }
                    }
                }
            }
        });
    }
};
