// @wix/flynt-rules/dist/catalog/breaking-changes/required-annotation-breaking-change/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
const annotations_utils_1 = require("../../annotations-utils");
const fqdn_utils_1 = require("../../fqdn-utils");
function isParentReachable(optionValue, startFrom) {
    const [_, ...paths] = optionValue.split('.');
    paths.pop();
    if (paths.length === 0) {
        return true;
    }
    else {
        return !!(0, annotations_utils_1.findFieldByPath)(startFrom, paths.join('.'));
    }
}
exports.rule = {
    description: '`required` annotation is a breaking change',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/protobuf#platformization-guidelines_protobuf_required-fields',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visitModified({
            visitMethod(method, prevMethod) {
                const prevRequired = prevMethod.options.getTypedOption(wix_protos_1.wix.api.required).map(r => r.value);
                const currentRequired = method.options.getTypedOption(wix_protos_1.wix.api.required).map(r => r.value);
                const isSpi = (0, fqdn_utils_1.isSpiService)(method.parent);
                const requiredMessage = isSpi ? prevMethod.response : prevMethod.request;
                const requiredPrefix = requiredMessage.name + '.';
                currentRequired.forEach(requiredField => {
                    if (requiredField.startsWith(requiredPrefix) &&
                        !prevRequired.includes(requiredField) &&
                        isParentReachable(requiredField, requiredMessage)) {
                        ctx.report({
                            errorAt: method,
                            message: `introducing of the 'wix.api.required' annotation for the '${requiredField}' is a breaking change`
                        });
                    }
                });
            }
        });
    }
};
