// @wix/flynt-rules/dist/catalog/errors/duplicate-rule-name/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'Custom validation errors in a method must have unique rule names',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2021-12-20'),
    run(ctx) {
        ctx.visit({
            visitMethod(method) {
                const errors = method.options
                    .getTypedOption(wix_protos_1.wix.api.custom_validation_error)
                    .filter(opt => opt.value.ruleName);
                if (errors.length > 0) {
                    const duplicateError = (0, fqdn_utils_1.findDuplicate)(errors, 'ruleName');
                    ctx.assert(!duplicateError, {
                        errorAt: duplicateError,
                        message: `the method '${method.name}' has a custom validation error with non unique rule name '${duplicateError?.value.ruleName}'`
                    });
                }
            }
        });
    }
};
