// @wix/flynt-rules/dist/catalog/errors/duplicate-application-code/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'Errors in a method must have unique application code',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2021-12-20'),
    run(ctx) {
        ctx.visit({
            visitMethod(method) {
                const errors = method.options
                    .getTypedOption(wix_protos_1.wix.api.error)
                    .filter(opt => opt.value.applicationCode);
                if (errors.length > 0) {
                    const duplicateApplicationError = (0, fqdn_utils_1.findDuplicate)(errors, 'applicationCode');
                    ctx.assert(!duplicateApplicationError, {
                        errorAt: duplicateApplicationError,
                        message: `the method '${method.name}' has an error with non unique application code '${duplicateApplicationError?.value.applicationCode}'`
                    });
                }
            }
        });
    }
};
