// @wix/locale-dataset-javascript/dist/test/builders/AddressPartsBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allAddressParts = exports.AddressPartsBuilder = void 0;
var AddressPartsBuilder = /** @class */ (function () {
    function AddressPartsBuilder() {
    }
    AddressPartsBuilder.prototype.withCountry = function (country) {
        this.country = country;
        return this;
    };
    AddressPartsBuilder.prototype.withSubdivision = function (subdivision) {
        this.subdivision = subdivision;
        return this;
    };
    AddressPartsBuilder.prototype.withAddressLine1 = function (addressLine1) {
        this.addressLine1 = addressLine1;
        return this;
    };
    AddressPartsBuilder.prototype.withAddressLine2 = function (addressLine2) {
        this.addressLine2 = addressLine2;
        return this;
    };
    AddressPartsBuilder.prototype.withCity = function (city) {
        this.city = city;
        return this;
    };
    AddressPartsBuilder.prototype.withZipCode = function (zipCode) {
        this.zipCode = zipCode;
        return this;
    };
    AddressPartsBuilder.prototype.withFirstName = function (firstName) {
        this.firstName = firstName;
        return this;
    };
    AddressPartsBuilder.prototype.withLastName = function (lastName) {
        this.lastName = lastName;
        return this;
    };
    AddressPartsBuilder.prototype.withCompany = function (company) {
        this.company = company;
        return this;
    };
    AddressPartsBuilder.prototype.withPhone = function (phone) {
        this.phone = phone;
        return this;
    };
    return AddressPartsBuilder;
}());
exports.AddressPartsBuilder = AddressPartsBuilder;
function allAddressParts() {
    return new AddressPartsBuilder()
        .withFirstName('Rina')
        .withLastName('Abadi')
        .withCountry('United States')
        .withSubdivision('NY')
        .withCity('New York')
        .withZipCode('1234')
        .withPhone('1-900-000-0000')
        .withAddressLine1('235 West 23rd Street')
        .withAddressLine2('8th Floor')
        .withCompany('Wix Lounge');
}
exports.allAddressParts = allAddressParts;
