// @wix/flynt-rules/dist/catalog/test-rules/test-rule-proto-root/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const rule = {
    description: 'Test rule for validating access to non proto files under mainProtoFolder',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if (service.name.includes('proto_xsyaiqjxksxdhsanpklg')) {
                    const docPath = service.sourceCode.leadingComments.split('\n')[0].trim();
                    if (docPath) {
                        ctx.assert(!fs_1.default.existsSync(path_1.default.join(ctx.protoDir, docPath)), {
                            errorAt: service,
                            message: `found a file under ${path_1.default.resolve(path_1.default.join(ctx.protoDir, docPath))}`,
                        });
                    }
                }
            }
        });
    }
};
exports.rule = rule;
