// @wix/flynt-rules/dist/catalog/fqdns/missing-main-entity-fqdn/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'A message pointed to by service must have the fqdn annotation',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/fqdn#platformization-guidelines_fqdn_annotate-the-main-entity-of-the-service',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2021-12-20'),
    run(ctx) {
        ctx.visit({
            visitService(service) {
                const mainEntityMessage = (0, fqdn_utils_1.resolveMainEntityMessage)(service);
                if (mainEntityMessage) {
                    const messageOptions = mainEntityMessage.options.getTypedOption(wix_protos_1.wix.api.entity);
                    ctx.assert(messageOptions && messageOptions.value.fqdn, {
                        errorAt: mainEntityMessage,
                        message: `the message '${(0, fqdn_utils_1.toLocalFqn)(mainEntityMessage.fqn)}' pointed to by the service '${(0, fqdn_utils_1.toLocalFqn)(service.fqn)}' must have the 'fqdn' annotation`
                    });
                }
            }
        });
    }
};
