// @wix/proto-syntax-tree/lib/syntax-tree/descriptor-paths.ts
//These are constants for the field numbers in proto of each field in FileDescriptorProto, ServiceDescriptorProto etc.
//They are uses in order to associate source code info with the appropriate object in the syntax tree
//See the documentation for descriptor.proto SourceCodeInfo for more info
//https://github.com/protocolbuffers/protobuf/blob/3f5fc4df1de8e12b2235c3006593e22d6993c3f5/src/google/protobuf/descriptor.proto#L754

export enum FileDescriptorPath {
  package = 2,
  message = 4,
  enum = 5,
  service = 6,
  extensions = 7,
  options = 8,
  syntax = 12,
}

export enum ServiceDescriptorPath {
  name = 1,
  method = 2,
  options = 3,
}

export enum MethodDescriptorPath {
  name = 1,
  input = 2,
  output = 3,
  options = 4,
}

export enum MessageDescriptorPath {
  name = 1,
  field = 2,
  message = 3,
  enum = 4,
  options = 7,
  oneof = 8,
}

export enum EnumDescriptorPath {
  name = 1,
  value = 2,
  options = 3,
}

export enum EnumValueDescriptorPath {
  name = 1,
  number = 2,
  options = 3,
}

export enum FieldDescriptorPath {
  name = 1,
  number = 3,
  label = 4,
  type = 5,
  typeName = 6,
  options = 8,
}

export enum OneofDescriptorPath {
  name = 1,
  options = 2,
}
