// @wix/proto-syntax-tree/lib/syntax-tree/source-code-info.ts
import {SourceCodeContainer, SourceCodeInfo} from '../api';
import {Location} from '../descriptor';

export class SourceCodeInfoImpl implements SourceCodeInfo {
  constructor(location: Location, container: SourceCodeContainer, name?: string) {
    this.leadingComments = location.leadingComments;
    this.trailingComments = location.trailingComments;
    this.leadingDetachedComments = location.leadingDetachedCommentsList;
    const span = location.spanList;
    this.startLine = span[0] + 1;
    this.startColumn = span[1] + 1;
    if (span.length > 3) {
      this.endLine = span[2] + 1;
      this.endColumn = span[3] + 1;
    } else {
      this.endLine = this.startLine;
      this.endColumn = span[2] + 1;
    }
    this.container = container;
    if (name) {
      this.fqn = `${this.container.fqn}.(${name})`;
    } else {
      this.fqn = this.container.fqn;
    }
  }

  readonly fqn: string;
  readonly container: SourceCodeContainer;
  readonly endColumn: number;
  readonly endLine: number;
  readonly leadingComments: string;
  readonly leadingDetachedComments: string[];
  readonly startColumn: number;
  readonly startLine: number;
  readonly trailingComments: string;
}
