// @wix/flynt-rules/dist/catalog/deprecations/deprecate-method-before-delete/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
const annotations_1 = require("@wix/wix-protos/typed/wix/api/annotations");
const fqdn_utils_1 = require("../../fqdn-utils");
const knownGeneratedMethods = ['SayHello'];
exports.rule = {
    description: 'A method should be deprecated before deleting',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/versioning-&-deprecation#platformization-guidelines_versioning-&-deprecation_breaking-changes',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visitRemoved({
            shouldVisitService(service) {
                const serviceMaturity = service.options.getTypedOption(wix_protos_1.wix.api.service_maturity);
                const isAlphaMaturity = serviceMaturity?.value === annotations_1.Maturity.ALPHA;
                return !(0, fqdn_utils_1.isPrivateExposure)(service) && !isAlphaMaturity;
            },
            visitMethod(method, parent) {
                if (knownGeneratedMethods.includes(method.name)) {
                    return;
                }
                const descOptions = method.options.descOptions;
                ctx.assert(descOptions.options.deprecated, {
                    errorAt: parent,
                    message: `The method '${method.name}' should not be removed without deprecating it first`
                });
            }
        });
    }
};
