// @wix/flynt-rules/dist/catalog/errors/invalid-data-message/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'Error data must be a valid message',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/errors#platformization-guidelines_errors_declare-errors-in-proto',
    severity: flynt_engine_1.Severity.Warning,
    strict: true,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if ((0, fqdn_utils_1.isSpiService)(service)) {
                    return false;
                }
            },
            visitMethod(method) {
                const errors = method.options
                    .getTypedOption(wix_protos_1.wix.api.error)
                    .filter(opt => opt.value.data);
                for (const error of errors) {
                    const dataMessageFqn = error.value.data;
                    if (dataMessageFqn) {
                        ctx.assertTypeExists(dataMessageFqn, flynt_engine_1.TypeName.Message, {
                            errorAt: error,
                            message: 'wix.api.error data should be a valid message'
                        });
                    }
                }
            }
        });
    }
};
