// @wix/flynt-rules/dist/catalog/fqdns/missing-main-entity/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
exports.rule = {
    description: 'A service must point to an existing main entity',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/fqdn#platformization-guidelines_fqdn_annotate-the-service-as-related-to-that-entity',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2021-12-20'),
    run(ctx) {
        ctx.visit({
            visitService(service) {
                const mainEntityFqn = (0, fqdn_utils_1.resolveMainEntityFqn)(service);
                if (mainEntityFqn) {
                    ctx.assertTypeExists(mainEntityFqn, flynt_engine_1.TypeName.Message, {
                        errorAt: service,
                        message: `the message '${mainEntityFqn}' defined in the service '${service.fqn}' in the option 'wix.api.service_entity' must exist`
                    });
                }
            }
        });
    }
};
