// @wix/flynt-rules/dist/catalog/annotations/unwrap-array-value/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
const proto_syntax_tree_1 = require("@wix/proto-syntax-tree");
exports.rule = {
    description: 'Message with unwrap_array_value must have a single repeated field and nothing else and cannot be used as response or request',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/api-design-patterns#platformization-guidelines_api-design-patterns_partial-update-of-arrays',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitMethod(method) {
                validateRequestResponse(method.request);
                validateRequestResponse(method.response);
            },
            visitMessage(message) {
                if (message.options.getTypedOption(wix_protos_1.wix.api.unwrap_array_value)) {
                    ctx.assert(message.fields.length === 1 && message.fields[0].fieldType.kind === proto_syntax_tree_1.FieldKind.Repeated, {
                        errorAt: message,
                        message: 'Message with unwrap_array_value must have a single repeated field and nothing else'
                    });
                }
            }
        });
        function validateRequestResponse(message) {
            ctx.assert(!message.options.getTypedOption(wix_protos_1.wix.api.unwrap_array_value), {
                errorAt: message,
                message: 'Message with unwrap_array_value cannot be used as response or request'
            });
        }
    }
};
