// @wix/flynt-rules/dist/catalog/permissions/missing-permission/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'A method must have a permission option defined',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/authorization-&-permissions#platformization-guidelines_authorization-&-permissions_1-in-your-proto-files',
    run(ctx) {
        ctx.visit({
            visitService(service) {
                return !(0, fqdn_utils_1.isSpiService)(service);
            },
            visitMethod(method) {
                if ((0, fqdn_utils_1.isSubscription)(method)) {
                    return false;
                }
                const permission = method.options.getTypedOption(wix_protos_1.wix.api.permission);
                ctx.assert(permission && permission.value.name, {
                    errorAt: method,
                    message: `the method '${method.name}' is missing a 'wix.api.permission' option`
                });
            }
        });
    }
};
