// @wix/flynt-rules/dist/catalog/annotations/maturity-ia-not-allowed/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
const annotations_1 = require("@wix/wix-protos/typed/wix/api/annotations");
exports.rule = {
    description: 'IA Maturity is deprecated',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/protobuf#platformization-guidelines_protobuf_maturity',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2022-01-24'),
    run(ctx) {
        ctx.visit({
            visitService(service) {
                const maturity = service.options.getTypedOption(wix_protos_1.wix.api.service_maturity);
                if (maturity) {
                    ctx.assert(maturity.value !== annotations_1.Maturity.IA, {
                        errorAt: maturity,
                        message: `Deprecated maturity 'IA' used in service: '${service.name}'`
                    });
                }
            },
            visitMethod(method) {
                const maturity = method.options.getTypedOption(wix_protos_1.wix.api.maturity);
                if (maturity) {
                    ctx.assert(maturity.value !== annotations_1.Maturity.IA, {
                        errorAt: maturity,
                        message: `Deprecated maturity 'IA' used in method: '${method.name}'`
                    });
                }
            },
            visitField(field) {
                const maturity = field.options.getTypedOption(wix_protos_1.wix.api.field_maturity);
                if (maturity) {
                    ctx.assert(maturity.value !== annotations_1.Maturity.IA, {
                        errorAt: maturity,
                        message: `Deprecated maturity 'IA' used in field: '${field.name}'`
                    });
                }
            }
        });
    }
};
