// @wix/flynt-rules/dist/catalog/spi/no-readonly-immutable-spi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'No read-only/immutable annotations in SPIs',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if (!(0, fqdn_utils_1.isSpiService)(service)) {
                    return false;
                }
            },
            visitMethod(method) {
                ctx.visitFrom(method.request, {
                    visitField(field) {
                        ctx.assert(field.options.getTypedOption(wix_protos_1.wix.api.readOnly) === undefined, {
                            errorAt: field,
                            message: 'Read-only annotation is not allowed for fields in an SPI'
                        });
                        ctx.assert(field.options.getTypedOption(wix_protos_1.wix.api.immutable) === undefined, {
                            errorAt: field,
                            message: 'Immutable annotation is not allowed for fields in an SPI'
                        });
                    }
                });
            }
        });
    }
};
