// @wix/proto-syntax-tree/dist/lib/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionType = exports.Primitives = exports.FieldKind = exports.ProtoFolderType = exports.TypeName = void 0;
var TypeName;
(function (TypeName) {
    TypeName["File"] = "File";
    TypeName["Service"] = "Service";
    TypeName["Method"] = "Method";
    TypeName["Message"] = "Message";
    TypeName["Field"] = "Field";
    TypeName["Oneof"] = "Oneof";
    TypeName["Enum"] = "Enum";
    TypeName["EnumValue"] = "EnumValue";
    TypeName["Extension"] = "Extension";
})(TypeName = exports.TypeName || (exports.TypeName = {}));
var ProtoFolderType;
(function (ProtoFolderType) {
    ProtoFolderType[ProtoFolderType["MainProtoFolder"] = 0] = "MainProtoFolder";
    ProtoFolderType[ProtoFolderType["IncludeProtoFolder"] = 1] = "IncludeProtoFolder";
})(ProtoFolderType = exports.ProtoFolderType || (exports.ProtoFolderType = {}));
var FieldKind;
(function (FieldKind) {
    FieldKind["Primitive"] = "Primitive";
    FieldKind["Wrapper"] = "Wrapper";
    FieldKind["Enum"] = "Enum";
    FieldKind["Message"] = "Message";
    FieldKind["Repeated"] = "Repeated";
    FieldKind["Map"] = "Map";
})(FieldKind = exports.FieldKind || (exports.FieldKind = {}));
var Primitives;
(function (Primitives) {
    Primitives[Primitives["DOUBLE"] = 1] = "DOUBLE";
    Primitives[Primitives["FLOAT"] = 2] = "FLOAT";
    Primitives[Primitives["INT64"] = 3] = "INT64";
    Primitives[Primitives["UINT64"] = 4] = "UINT64";
    Primitives[Primitives["INT32"] = 5] = "INT32";
    Primitives[Primitives["FIXED64"] = 6] = "FIXED64";
    Primitives[Primitives["FIXED32"] = 7] = "FIXED32";
    Primitives[Primitives["BOOL"] = 8] = "BOOL";
    Primitives[Primitives["STRING"] = 9] = "STRING";
    Primitives[Primitives["BYTES"] = 12] = "BYTES";
    Primitives[Primitives["UINT32"] = 13] = "UINT32";
    Primitives[Primitives["SFIXED32"] = 15] = "SFIXED32";
    Primitives[Primitives["SFIXED64"] = 16] = "SFIXED64";
    Primitives[Primitives["SINT32"] = 17] = "SINT32";
    Primitives[Primitives["SINT64"] = 18] = "SINT64";
})(Primitives = exports.Primitives || (exports.Primitives = {}));
var OptionType;
(function (OptionType) {
    OptionType[OptionType["FileOptions"] = 1] = "FileOptions";
    OptionType[OptionType["ServiceOptions"] = 2] = "ServiceOptions";
    OptionType[OptionType["MethodOptions"] = 3] = "MethodOptions";
    OptionType[OptionType["MessageOptions"] = 4] = "MessageOptions";
    OptionType[OptionType["EnumOptions"] = 5] = "EnumOptions";
    OptionType[OptionType["EnumValueOptions"] = 6] = "EnumValueOptions";
    OptionType[OptionType["FieldOptions"] = 7] = "FieldOptions";
    OptionType[OptionType["OneofOptions"] = 8] = "OneofOptions";
})(OptionType = exports.OptionType || (exports.OptionType = {}));
