// @wix/flynt-rules/dist/catalog/field-types/unsigned-integers-not-allowed/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const proto_syntax_tree_1 = require("@wix/proto-syntax-tree");
const rule = {
    description: 'Unsigned integers are not allowed',
    breakIfNew: true,
    run(ctx) {
        ctx.visit({
            visitField(field) {
                validateFieldType(field.fieldType, field);
            }
        });
        function validateFieldType(fieldType, field) {
            switch (fieldType.kind) {
                case flynt_engine_1.FieldKind.Primitive:
                case flynt_engine_1.FieldKind.Wrapper: {
                    assertNotUnsigned(fieldType.type, field);
                    break;
                }
                case flynt_engine_1.FieldKind.Map: {
                    assertNotUnsigned(fieldType.keyType, field);
                    validateFieldType(fieldType.valueType, field);
                    break;
                }
            }
        }
        function assertNotUnsigned(type, field) {
            ctx.assert(type !== proto_syntax_tree_1.Primitives.UINT32 && type !== proto_syntax_tree_1.Primitives.UINT64, {
                errorAt: field,
                message: `${field.name} has type of unsigned integer, unsigned integers are not allowed`
            });
        }
    }
};
exports.rule = rule;
