// @wix/flynt-rules/dist/catalog/events/missing-segment/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = exports.isFqnInLegacyCluster = exports.isDomainEventFqdnInLegacyCluster = exports.domainEventTopic = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
const callback_1 = require("@wix/wix-protos/typed/wix/api/callback");
const { legacyTopics } = require('@wix/greyhound-legacy-topics');
const DOMAIN_EVENTS_PREFIX = 'domain_events_';
const FQNS_ALLOWED_TO_HAVE_NO_SEGMENT = new Set([
    'com.wixpress.contacts.core.api.v4.ContactBulkJob',
    'com.wixpress.users.User',
    'com.wixpress.vi.gdpr.Product',
    'wix.proto.registry.ProtoFilesSet',
    'com.wix.bookings.calendar.api.v1.CalendarFeedEvent',
    'com.wix.corechat.auth.Token',
    'com.wix.fed.settings.AppSettings',
    'com.wix.pricingplans.benefits.api.v2.Benefit',
    'com.wix.pricingplans.benefits.api.v2.MemberBenefit',
    'com.wixpress.achievements.api.v1.ApprovalRequest',
    'com.wixpress.achievements.api.v1.Category',
    'com.wixpress.achievements.api.v1.Challenge',
    'com.wixpress.achievements.api.v1.GetChallengeStatsResponse',
    'com.wixpress.amitis.wishlist.Member',
    'com.wixpress.analytics.ng.api.v1.Dashboard',
    'com.wixpress.analytics.ng.api.v1.Data',
    'com.wixpress.analytics.ng.api.v1.Forecast',
    'com.wixpress.analytics.ng.api.v1.Insight',
    'com.wixpress.buildup.spi.CrossSellOrder',
    'com.wixpress.cloud.secrets_vault.api.v1.Secret',
    'com.wixpress.coreservices.velo.autogen.definitions.v1.edm.definition.EdmDefinition',
    'com.wixpress.critique.configuration.Configuration',
    'com.wixpress.crm.financial.invoices.api.v2.Invoice',
    'com.wixpress.crm.financial.invoices.api.v2.InvoiceSettings',
    'com.wixpress.crm.financial.settings.api.v2.DesignSettings',
    'com.wixpress.crm.financial.settings.api.v2.Product',
    'com.wixpress.crm.financial.settings.api.v2.Tax',
    'com.wixpress.devcenter.appcoupons.Coupon',
    'com.wixpress.document.api.Component',
    'com.wixpress.document.api.Page',
    'com.wixpress.document.api.Settings',
    'com.wixpress.ecommerce.shipping.options.v1.Carrier',
    'com.wixpress.editorlocalizer.TranslationProject',
    'com.wixpress.experts.loyalty.api.v2.LoyaltyData',
    'com.wixpress.experts.proposals.ProposalTemplate',
    'com.wixpress.falcon.build_server_projects.api.Project',
    'com.wixpress.fed.depsgraph.Package',
    'com.wixpress.forms.submissions.view.SubmissionView',
    'com.wixpress.gift.cards.proxy.Transaction',
    'com.wixpress.gretak.comments.Comment',
    'com.wixpress.ims.api.Issue',
    'com.wixpress.keynote.questions.Question',
    'com.wixpress.localizationadapters.EditorComponent',
    'com.wixpress.loyalty.account.Transaction',
    'com.wixpress.members.defaultprivacy.DefaultPrivacy',
    'com.wixpress.members.memberroles.MemberRoleDefinition',
    'com.wixpress.members.reports.Report',
    'com.wixpress.membership.v1.orders.comments.OrderComment',
    'com.wixpress.membership.v1.tax.TaxConfig',
    'com.wixpress.membership.v2.settings.Settings',
    'com.wixpress.multilingual.api.SettingsData',
    'com.wixpress.npm.blog.revision.history.v1.PostRevision',
    'com.wixpress.npm.communities.platformized.blog.v3.RichContent',
    'com.wixpress.npm.dummy.EchoPayload',
    'com.wixpress.npm.dummy.EchoPayload2',
    'com.wixpress.payments.bo_activities.v1.BoActivity',
    'com.wixpress.payments.bo_risk.v1.HighRiskSiteRecord',
    'com.wixpress.payments.bo_risk.v1.MotoSetting',
    'com.wixpress.pos.release.tool.api.ReleaseCandidate',
    'com.wixpress.premium.licenser.License',
    'com.wixpress.premium.provisioning.ProvisioningSet',
    'com.wixpress.promote.campaigns.manager.v1.Account',
    'com.wixpress.promote.campaigns.manager.v1.CreativeTypeItem',
    'com.wixpress.promote.campaigns.manager.v1.FacebookSetup',
    'com.wixpress.promote.fbe.service.v1.FacebookSetup',
    'com.wixpress.promote.promote_insights_public.Post',
    'com.wixpress.promote.promote_insights_public.TargetedActionableIntent',
    'com.wixpress.promote.publisher.v1.Account',
    'com.wixpress.promote.publisher.v1.Item',
    'com.wixpress.promote.seo.server.v1.Achievement',
    'com.wixpress.promote.seo.server.v1.BusinessData',
    'com.wixpress.promote_campaigns_manager.v2.Campaign',
    'com.wixpress.promote_campaigns_manager.v2.CampaignEligibilityResponse',
    'com.wixpress.promote_campaigns_manager.v2.ExposureInfo',
    'com.wixpress.promote_campaigns_manager.v2.UserPackage',
    'com.wixpress.restaurants.localdelivery.order.Order',
    'com.wixpress.sitelist.api.Site',
    'com.wixpress.teamwork.teamworkorchestrator.RepoUrlMapping',
    'com.wixpress.troubleshooting.something.Message',
    'com.wixpress.vi.usermanager.api.apps.v1.PremiumApp',
    'com.wixpress.vi.usermanager.api.backofficeusers.v1.BoUserInfo',
    'com.wixpress.vi.usermanager.api.backofficeusers.v1.Permission',
    'com.wixpress.vi.usermanager.api.billing.v1.Invoice',
    'com.wixpress.vi.usermanager.api.billing.v1.PaymentMethod',
    'com.wixpress.vi.usermanager.api.domains.v1.TransferredInDomain',
    'com.wixpress.vi.usermanager.api.editor.v1.SiteRevision',
    'com.wixpress.vi.usermanager.api.mailboxes.v1.GoogleMailboxAccountInfo',
    'com.wixpress.vi.usermanager.api.mailboxes.v1.MailboxPackage',
    'com.wixpress.vi.usermanager.api.plans.v1.License',
    'com.wixpress.vi.usermanager.api.plans.v1.Plan',
    'com.wixpress.vi.usermanager.api.sites.v1.Album',
    'com.wixpress.vi.usermanager.api.sites.v1.DeletedSite',
    'com.wixpress.vi.usermanager.api.sites.v1.Site',
    'com.wixpress.vi.usermanager.api.users.v1.User',
    'com.wixpress.vi.usermanager.api.vouchers.v1.Voucher',
    'com.wixpress.virtual.numbers.communication.records.SmsRecord',
    'com.wixpress.virtual.numbers.manager.api.v1.AgentInfo',
    'com.wixpress.virtual.numbers.manager.api.v1.AssignedAgent',
    'com.wixpress.virtual.numbers.manager.api.v1.CallDetailRecord',
    'com.wixpress.virtual.numbers.manager.api.v1.CommunicationChannel',
    'com.wixpress.virtual.numbers.manager.api.v1.PhoneNumber',
    'com.wixpress.virtual.numbers.manager.api.v1.UsageRecord',
    'com.wixpress.virtual.numbers.manager.api.v1.VirtualNumberAccountInfo',
    'com.wixpress.virtual.numbers.manager.api.v1.Voicemail',
    'com.wixpress.workflow.api.v1.Phase',
    'com.wixpress.workflow.api.v1.Workflow',
    'com.wixpress.workflow.api.v1.WorkflowInfo',
    'test.dbchauffeur.test.pb.TestMessage',
    'wix.analytics_ng.dashboards.v2.Dashboard',
    'wix.analytics_ng.data.v2.Contact',
    'wix.analytics_ng.data.v2.Overview',
    'wix.analytics_ng.data.v2.Stats',
    'wix.analytics_ng.forecasts.v2.Forecast',
    'wix.analytics_ng.insights.v2.Insight',
    'wix.analytics_ng.subscriptions.v2.Subscription',
    'wix.ascend.campaign.v1.service.Activity',
    'wix.ascend.campaign.v1.service.Campaign',
    'wix.ascend.promotional_site_assets_generation.v1.service.Assets',
    'wix.bookmarks.Bookmark',
    'wix.bookmarks.BookmarkFolder',
    'wix.cashier.v2.settings.Merchant',
    'wix.cashier.v2.settings.PaymentMethodType',
    'wix.cashier.v2.settings.PaymentProvider',
    'wix.cashier.v2.settings.ProviderAccount',
    'wix.coreservices.businessschema.v1.Appendix',
    'wix.coreservices.metroinspector.v1.products.Product',
    'wix.dam.api.v1.Project',
    'wix.dbchauffeur.test.AllTypes.InnerMessage',
    'wix.dealerAnalytics.api.v1.AudienceAnalytics',
    'wix.dealerAnalytics.api.v1.OfferAnalytics',
    'wix.dealerBackoffice.api.v1.Audience',
    'wix.dealerBackoffice.api.v1.Offer',
    'wix.dealerOfferEvents.api.v1.OfferEvent',
    'wix.dealerOffersServing.api.v1.Offer',
    'wix.diners.v1.Restaurant',
    'wix.discovery.ingester.v1.DiscoverableContent',
    'wix.docs.v1.Portal',
    'wix.ecommerce.subscription.option.api.v1.SubscriptionOption',
    'wix.events.Event',
    'wix.events.categories.Category',
    'wix.events.rsvp.Rsvp',
    'wix.events.ticketing.Order',
    'wix.events.ticketing.Ticket',
    'wix.events.ticketing.TicketDefinition',
    'wix.faq.v1.Category',
    'wix.faq.v1.QuestionEntry',
    'wix.filesharing.api.v1.libraryitems.LibraryItem',
    'wix.filesharing.api.v1.settings.AppSettings',
    'wix.fulfillment.api.v1.Fulfiller',
    'wix.helpcenter.HelpCenterSession',
    'wix.livevideo.ChatMessage',
    'wix.livevideo.LiveSession',
    'wix.livevideo.Participant',
    'wix.livevideo.Reaction',
    'wix.livevideo.VideoRecording',
    'wix.livevideo.VideoStream',
    'wix.payment.api.dashboard.Payment',
    'wix.payment.api.dashboard.account.SiteInfo',
    'wix.payments.bo_accounts.v4.BoAccount',
    'wix.payments.bo_balances.v4.BoBalanceRecord',
    'wix.payments.bo_payout_policies.v4.BoPayoutPolicy',
    'wix.payments.bo_payouts.v4.BoPayout',
    'wix.payments.card_tokens.v1.CardToken',
    'wix.payments.dispute_cases.v1.DisputeCase',
    'wix.payments.internal.accounts.InternalAccount',
    'wix.payments.internal.balances.InternalBalanceRecord',
    'wix.payments.v1.topups.Topup',
    'wix.payments.v3.payouts.Payout',
    'wix.payments.v4.balances.BalanceRecord',
    'wix.payments.v4.mcc.MerchantCategoryCodes',
    'wix.payments.v4.payouts.Payout',
    'wix.payments.v5.balances.BalanceRecord',
    'wix.reactions.IdentityReaction',
    'wix.reactions.ResourceReactionsSummary',
    'wix.release_manager.v1.RevisionsExposure',
    'wix.social.groups.api.v1.EmailInviteRequest',
    'wix.social.groups.api.v1.GetActivityStatsResponse',
    'wix.social.groups.api.v1.GroupApp',
    'wix.social.groups.api.v1.GroupEvent',
    'wix.social.groups.api.v1.GroupMembership',
    'wix.social.groups.api.v1.GroupResponse',
    'wix.social.groups.api.v1.GroupRole',
    'wix.social.groups.api.v1.GroupsInternal',
    'wix.social.groups.api.v1.NotificationSettings',
    'wix.social.groups.api.v1.Rule',
    'wix.social.groups.api.v2.AutoInvite',
    'wix.social.groups.api.v2.GroupsReady',
    'wix.social.groups.api.v2.MembershipQuestion',
    'wix.social.groups.api.v2.ResetUpdates',
    'wix.topics.Topic',
    'wix.velo.apps.v1.App',
    'wix.velo.spi.v1.SpiConnectionStatus',
    'wix.vod.api.v3.ChannelVideo',
    'wix.wixpay.payout.api.internal.InternalPayout',
    'wix.wixpay.transaction.api.internal.ChargeSimulationRule',
    'wix.wixpay.transaction.api.internal.InternalTransaction',
    'com.wix.crm.FormPage',
    'com.wixpress.ci.loki.api.LokiAdmin',
    'com.wixpress.ci.loki.api.Update',
    'com.wixpress.juliusl.comments.serivce.Comment',
    'com.wixpress.promote.search.engine.service.v1.SearchEngineData',
    'com.wixpress.restaurants.settings.Settings',
    'com.wixpress.vi.usermanager.api.domains.v1.Domain',
    'com.wixpress.vi.usermanager.api.domains.v1.WixDomainRegistration',
    'wix.discovery.ingester.v1.DiscoverySettings',
    'wix.templates.api.backoffice.v1.Book',
    'wix.templates.api.backoffice.v1.Category',
    'wix.templates.api.backoffice.v1.Tag',
    'wix.templates.api.backoffice.v1.Template',
    'wix.templates.api.backoffice.v2.BookAbTest',
    'wix.templates.api.backoffice.v2.CategoryAbTest',
    'wix.templates.api.backoffice.v2.TemplateAbTest',
    'com.wixpress.subscriptions.api.v1.Subscription',
]);
function domainEventTopic(fqdn) {
    return DOMAIN_EVENTS_PREFIX + fqdn;
}
exports.domainEventTopic = domainEventTopic;
function isDomainEventFqdnInLegacyCluster(fqdn) {
    return legacyTopics.includes(domainEventTopic(fqdn));
}
exports.isDomainEventFqdnInLegacyCluster = isDomainEventFqdnInLegacyCluster;
function isFqnInLegacyCluster(fqn) {
    return FQNS_ALLOWED_TO_HAVE_NO_SEGMENT.has(fqn);
}
exports.isFqnInLegacyCluster = isFqnInLegacyCluster;
exports.rule = {
    description: 'A wix.api.entity must have the segment field specified',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/events#platformization-guidelines_events_12-choose-the-right-segment',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitMessage(mainEntityMessage) {
                const mainFqn = (0, fqdn_utils_1.toLocalFqn)(mainEntityMessage.fqn);
                if (isFqnInLegacyCluster(mainFqn)) {
                    return false;
                }
                const entityOption = mainEntityMessage.options.getTypedOption(wix_protos_1.wix.api.entity);
                if (entityOption) {
                    if (entityOption.value.fqdn) {
                        const topicName = entityOption.value.oldFqdnForBackwardsCompatibility ?? entityOption.value.fqdn;
                        ctx.assert((entityOption.value.segment !== undefined && entityOption.value.segment !== callback_1.Segment.Segment.UNSPECIFIED) ||
                            isDomainEventFqdnInLegacyCluster(topicName) ||
                            (0, fqdn_utils_1.isSpiFqdn)(topicName), {
                            errorAt: entityOption,
                            message: `A 'segment' field for the fqdn '${entityOption.value.fqdn}' must be specified`
                        });
                    }
                }
            },
        });
    }
};
