// @wix/wix-measured/lib/wix-measured-registry.js
const sanitize = require('./sanitize'),
  assert = require('assert');

module.exports = class WixMeasuredRegistry {
  constructor({prefix, meters = {}, hists = {}, gauges = {}}) {
    assert(prefix, 'prefix is mandatory');
    this._prefix = prefix;
    this._meters = meters;
    this._gauges = gauges;
    this._hists = hists;
  }

  get prefix() {
    return this._prefix;
  }

  get meters() {
    return this._meters;
  }

  get gauges() {
    return this._gauges;
  }

  get hists() {
    return this._hists;
  }

  addMeter(key, name, createMeter) {
    return this._addMetricTo(this._meters, key, name, createMeter);
  }

  addGauge(key, name, createGauge) {
    return this._addMetricTo(this._gauges, key, name, createGauge, false /* reuseExisting */);
  }

  addHist(key, name, createHist) {
    return this._addMetricTo(this._hists, key, name, createHist);
  }

  forCollection(key, name) {
    return new WixMeasuredRegistry({
      prefix: this._metricKey(key, name),
      meters: this._meters,
      hists: this._hists,
      gauges: this._gauges});
  }

  _addMetricTo(collection, key, name, createMetric, reuseExisting = true) {
    assert(typeof createMetric === 'function', 'metric creation function is mandatory');
    
    collection[this._metricKey(key, name)] = (reuseExisting && collection[this._metricKey(key, name)]) || createMetric();
    return collection[this._metricKey(key, name)];
  }

  _metricKey(key, name) {
    assert(key && typeof key === 'string', 'key must be a string and is mandatory');
    assert(name && typeof name === 'string', 'name must be a string and is mandatory');

    return [this._prefix, `${sanitize(key)}=${sanitize(name)}`].join('.');
  }
};
