// @wix/flynt-rules/dist/catalog/field-validations/field-cannot-be-immutable-and-readonly/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'A field cannot be both immutable and readOnly',
    moreInfoUrl: 'https://github.com/wix-private/server-infra/blob/master/framework/docs/adr/adr-018-readonly-and-immutable.md',
    breakIfNew: true,
    run(ctx) {
        ctx.visit({
            visitField(field) {
                const readOnlyOption = field.options.getTypedOption(wix_protos_1.wix.api.readOnly);
                const immutableOption = field.options.getTypedOption(wix_protos_1.wix.api.immutable);
                const readOnlyAndImmutable = immutableOption && immutableOption.value && readOnlyOption && readOnlyOption.value;
                ctx.assert(!readOnlyAndImmutable, {
                    errorAt: field,
                    message: `The field '${field.name}' cannot have both '(wix.api.readOnly) = true' and '(wix.api.immutable)=true'`
                });
            }
        });
    }
};
