// @wix/flynt-rules/dist/catalog/breaking-changes/removing-required-annotation-must-sdk-signature/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
const fqdn_utils_1 = require("../../fqdn-utils");
exports.rule = {
    description: 'the `required` annotation cannot be removed unless it is marked as sdk signature annotation',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/edms#platformization-guidelines_edms_override-with-sdk-signature-option',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visitModified({
            shouldVisitService(service) {
                return (0, fqdn_utils_1.isServiceRelevantForAutoGeneratedClient)(service);
            },
            visitMethod(method, prevMethod) {
                if ((0, fqdn_utils_1.isMethodRelevantForAutoGeneratedClient)(method)) {
                    const prevRequired = prevMethod.options.getTypedOption(wix_protos_1.wix.api.required).map(r => r.value);
                    const currentRequired = method.options.getTypedOption(wix_protos_1.wix.api.required).map(r => r.value);
                    const sdkSignature = method.options.getTypedOption(wix_protos_1.wix.sdk.signature)[0];
                    const signatureParams = (sdkSignature?.value.params || []).map(field => method.request.name + '.' + field);
                    prevRequired.forEach(required => {
                        if (!currentRequired.includes(required) && !signatureParams.includes(required)) {
                            ctx.report({
                                errorAt: method,
                                message: `the 'required' annotation '${required}' cannot be removed unless the '${required.substring(method.request.name.length + 1)}' field is marked as '.wix.sdk.signature'`
                            });
                        }
                    });
                }
            }
        });
    }
};
