// @wix/flynt-rules/dist/catalog/field-validations/fields-order-change-not-allowed/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
exports.rule = {
    description: 'Changing order of fields in requests is not allowed',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/edms#platformization-guidelines_edms_method-signature-setup',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if (!(0, fqdn_utils_1.isServiceRelevantForAutoGeneratedClient)(service)) {
                    return false;
                }
            },
            visitMethod(method) {
                if ((0, fqdn_utils_1.isMethodRelevantForAutoGeneratedClient)(method)) {
                    const removedCounts = new Map();
                    ctx.visitModifiedFrom(method.request, {
                        visitMessage(message, prevMessage) {
                            let removedCount = 0;
                            for (let prev = 0; prev < prevMessage.fields.length; prev++) {
                                const prevField = prevMessage.fields[prev];
                                if (message.fields.findIndex(field => field.name === prevField.name) === -1) {
                                    removedCount++;
                                }
                                removedCounts.set(prevField.name, removedCount);
                            }
                        },
                        visitField(field, prevField) {
                            const currentIndex = field.parent.fields.indexOf(field);
                            const prevIndex = prevField.parent.fields.indexOf(prevField);
                            const removedCount = removedCounts.get(prevField.name);
                            ctx.assert(currentIndex + removedCount === prevIndex, {
                                errorAt: field,
                                message: `Changing order of fields in requests is not allowed. Field ${field.name} order was changed from ${prevIndex} to ${currentIndex}`
                            });
                        }
                    });
                }
            }
        });
    }
};
