// @wix/flynt-rules/dist/catalog/events/no-events-callbacks-spi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
function findCallback(method) {
    return method.options.getTypedOption(wix_protos_1.wix.api.callback).find(opt => opt);
}
function findEvent(service) {
    return service.options.getTypedOption(wix_protos_1.wix.api.event).find(opt => opt);
}
exports.rule = {
    description: 'Events and callbacks are not supported in SPI services',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if (!(0, fqdn_utils_1.isSpiService)(service)) {
                    return false;
                }
                const event = findEvent(service);
                ctx.reportIf(event, {
                    errorAt: event,
                    message: `the service '${service.name}' cannot have 'wix.api.event' because it is an SPI service`
                });
            },
            visitMethod(method) {
                const callback = findCallback(method);
                ctx.reportIf(callback, {
                    errorAt: callback,
                    message: `the method '${method.name}' cannot have 'wix.api.callback' because '${method.parent.name}' is an SPI service`
                });
            }
        });
    }
};
