// @wix/flynt-engine/test/test-utils.ts
/* eslint-disable prefer-rest-params */
import {Enum, EnumValue, Extension, Field, Message, Method, Oneof, ProtoFile, Service, Type} from '@wix/proto-syntax-tree';
import {ShouldVisit, StatefulVisitor} from '../lib/stateful/rule-api';

export class StatefulLoggingVisitor implements StatefulVisitor, ShouldVisit {
  visited: [string, string?][] = [];

  private log(...args: any) {
    if (args.length === 1) {
      this.visited.push([`${args[0].fqn} [${args[0].typeName}]`]);
    } else if (args.length === 2) {
      this.visited.push([`${args[0].fqn} [${args[0].typeName}]`, args[1]?.fqn]);
    }
  }

  shouldVisitFile(_elem: ProtoFile): boolean {
    return true;
  }

  visitFile(_elem: ProtoFile, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  shouldVisitService(_elem: Service): boolean {
    return true;
  }

  visitService(_elem: Service, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  visitMethod(_elem: Method, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  shouldVisitMessage(_elem: Message): boolean {
    return true;
  }

  visitMessage(_elem: Message, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  shouldVisitEnum(_elem: Enum): boolean {
    return true;
  }

  visitEnum(_elem: Enum, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  visitEnumValue(_elem: EnumValue, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  visitField(_elem: Field, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  visitOneof(_elem: Oneof, _prevState?: Type | undefined) {
    this.log(...arguments);
  }

  visitExtension(_elem: Extension, _prevState?: Type | undefined) {
    this.log(...arguments);
  }
}
