// @wix/flynt-rules/dist/catalog/spi/fqdn-rule-for-spi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'wix.interfaces.* entities MUST only be referenced from SPI services and SPI services must contains only wix.interfaces.* entities',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface#platformization-guidelines_spi---service-provider-interface_spi---service-provider-interface',
    run(ctx) {
        ctx.visit({
            visitService(service) {
                function findFqdnOption() {
                    const spiOption = service.options.getTypedOption(wix_protos_1.wix.spi.service);
                    if (spiOption && spiOption.value.fqdn) {
                        return spiOption;
                    }
                }
                const fqdnOption = findFqdnOption();
                if (fqdnOption) {
                    const fqdn = fqdnOption.value.fqdn;
                    const isServiceSpi = (0, fqdn_utils_1.isSpiService)(service);
                    if ((0, fqdn_utils_1.isSpiFqdn)(fqdn)) {
                        ctx.assert(isServiceSpi, {
                            errorAt: fqdnOption,
                            message: 'fqdns that start with \'wix.interfaces.*\' should only be referenced by SPI services'
                        });
                    }
                    else {
                        ctx.assert(!isServiceSpi, {
                            errorAt: fqdnOption,
                            message: `The fqdn of '${(0, fqdn_utils_1.toLocalFqn)(service.name)}' must be in the form 'wix.interfaces.*' because it is an SPI`
                        });
                    }
                }
            }
        });
    }
};
