// @wix/flynt-rules/dist/catalog/breaking-changes/deleted-field-enum-reserved/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
exports.rule = {
    description: 'Deleted field/enum should be marked as reserved',
    moreInfoUrl: 'https://developers.google.com/protocol-buffers/docs/proto3#reserved',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visitRemoved({
            visitField(field, parent) {
                const isReservedInCurrent = parent.reserved.isReservedNumber(field.number);
                ctx.assert(isReservedInCurrent, {
                    errorAt: parent,
                    message: `the id '${field.number}' of the deleted field '${field.name}' should be marked as 'reserved' in the message '${parent.name}'`
                });
            },
            visitEnumValue(enumValue, parent) {
                const isReservedInCurrent = parent.reserved.isReservedNumber(enumValue.number);
                ctx.assert(isReservedInCurrent, {
                    errorAt: parent,
                    message: `the id '${enumValue.number}' of the deleted enum '${enumValue.name}' should be marked as 'reserved' in the enum '${parent.name}'`
                });
            }
        });
    }
};
