// @wix/wix-errors/lib/grpc-status.js
const makeEnum = require('./enum');

const GrpcStatus = makeEnum({
  OK: 0,
  CANCELLED: 1,
  UNKNOWN: 2,
  INVALID_ARGUMENT: 3,
  DEADLINE_EXCEEDED: 4,
  NOT_FOUND: 5,
  ALREADY_EXISTS: 6,
  PERMISSION_DENIED: 7,
  RESOURCE_EXHAUSTED: 8,
  FAILED_PRECONDITION: 9,
  ABORTED: 10,
  OUT_OF_RANGE: 11,
  UNIMPLEMENTED: 12,
  INTERNAL: 13,
  UNAVAILABLE: 14,
  DATA_LOSS: 15,
  UNAUTHENTICATED: 16
});

const HTTP_STATUS_MAPPING = {
  [GrpcStatus.OK]: 200,
  [GrpcStatus.CANCELLED]: 499,
  [GrpcStatus.UNKNOWN]: 500,
  [GrpcStatus.INVALID_ARGUMENT]: 400,
  [GrpcStatus.DEADLINE_EXCEEDED]: 504,
  [GrpcStatus.NOT_FOUND]: 404,
  [GrpcStatus.ALREADY_EXISTS]: 409,
  [GrpcStatus.PERMISSION_DENIED]: 403,
  [GrpcStatus.RESOURCE_EXHAUSTED]: 429,
  [GrpcStatus.FAILED_PRECONDITION]: 428,
  [GrpcStatus.ABORTED]: 409,
  [GrpcStatus.OUT_OF_RANGE]: 400,
  [GrpcStatus.UNIMPLEMENTED]: 501,
  [GrpcStatus.INTERNAL]: 500,
  [GrpcStatus.UNAVAILABLE]: 503,
  [GrpcStatus.DATA_LOSS]: 500,
  [GrpcStatus.UNAUTHENTICATED]: 401
};

Object.assign(GrpcStatus, {
  toHttpStatus(grpcStatus) {
    return HTTP_STATUS_MAPPING[grpcStatus];
  }
});

module.exports = {GrpcStatus};
