// @wix/flynt-rules/dist/catalog/spi/main-entity-not-allowed-in-spi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'Main entity is not allowed in SPI',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface#platformization-guidelines_spi---service-provider-interface_how-to-add-an-spi-component-in-dev-center',
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if ((0, fqdn_utils_1.isSpiService)(service)) {
                    const mainEntityOption = service.options.getTypedOption(wix_protos_1.wix.api.service_entity);
                    ctx.assert(!mainEntityOption, {
                        errorAt: mainEntityOption,
                        message: 'main entity should not be defined for SPI service, please use the \'fqdn\' property under the \'wix.spi.service\' annotation instead'
                    });
                }
            }
        });
    }
};
