// @wix/flynt-rules/dist/catalog/fqdns/legacy-fqdn-whitelisted/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
const invalid_message_fqdn_name_1 = require("../invalid-message-fqdn-name");
const whitelistOfLegacyFqdns = new Set([
    'wix.sm.v1.member.reset_password',
    'wix.sm.v1.member.status_changed',
    'com.wix.bookings.conferencing.ConferenceAccount',
    'com.wix.bookings.conferencing.v1.conference_account',
    'com.wix.ecom.orders.payments.v1.OrderTransactions',
    'com.wixpress.experts.feedback.Feedback',
    'com.wixpress.infra.sdl.Person',
    'com.wixpress.infra.sdl.dbtest.MyEntity',
    'com.wixpress.infra.sdl.playground.MyEntity',
    'com.wixpress.locations.Location',
    'com.wixpress.vi.gdpr.request',
    'com.wixpress.vi.gdpr.response',
    'com.wixpress.vi.gdpr.product',
    'wix.pricing_plans.plan',
    'com.wixpress.npm.communities.forum.api.category',
    'com.wixpress.npm.communities.forum.api.comment',
    'com.wixpress.npm.communities.forum.api.post',
    'com.wixpress.premium.plans.v1.PremiumTpaSubscriptionChanged',
    'com.wixpress.restaurants.Order',
    'com.wixpress.wishlisInfoFQDN',
    'com.wixpress.wixdocs.viewer',
    'serverless.deployer.artifact',
    'serverless.deployer.configuration',
    'serverless.deployer.deployable',
    'serverless.deployer.environment_variable',
    'serverless.deployer.secret',
    'wix.badges.badge',
    'wix.blog',
    'wix.blog.category',
    'wix.blog.post',
    'wix.collections.collection',
    'wix.comments.comment',
    'wix.communities.forum.category',
    'wix.communities.forum.comment',
    'wix.communities.forum.post',
    'wix.contacts.v4.contact-contacts-allocator-dead-letters',
    'wix.contacts.v4.export-task',
    'wix.contacts.v4.extended-field',
    'wix.contacts.v4.label-contacts-app-labels-cache-invalidator-dead-letters',
    'wix.crm.action_triggers.automation_template',
    'wix.crm.action_triggers.automation_template_group',
    'wix.crm.action_triggers.rule',
    'wix.crm.action_triggers.state',
    'wix.crm.action_triggers.trigger_configuration',
    'wix.crm.contacts.contact',
    'wix.crm.contacts.contact-contacts-app-merge-dead-letters',
    'wix.crm.contacts.extended-field',
    'wix.crm.contacts.field',
    'wix.crm.contacts.label',
    'wix.crm.financial.invoice.invoice_design_template',
    'wix.crm.financial.invoice.invoice_settings',
    'wix.crm.financial.invoice.invoice_template',
    'wix.crm.financial.settings.custom_field_definition',
    'wix.crm.financial.settings.custom_field_type',
    'wix.crm.financial.settings.design_settings',
    'wix.crm.virtual_numbers.communication_record_changed',
    'wix.dummy_topic',
    'wix.ecom.v1.checkout',
    'wix.ecom.v1.shipping.localdelivery.LocalDeliveryOption',
    'wix.ecom.v1.shipping.rule',
    'wix.ecommerce.coupons.v2.coupon-wix-code-bo-events-processor-dead-letters',
    'wix.ecommerce.gift_cards.proxy.v1.gift_card',
    'wix.editor.site_branches.v1.Branch',
    'wix.email.emailproxy',
    'wix.esb.configuration',
    'wix.experts.feedback',
    'wix.filesharing.api.v1.permissions.access_requested',
    'wix.filesharing.api.v1.permissions.permission_level_assigned',
    'wix.forms.v1.form_submitted',
    'wix.forum.category',
    'wix.forum.comment',
    'wix.forum.comment-wix-code-bo-events-processor-dead-letters',
    'wix.forum.post',
    'wix.ims.v1.issue_tag',
    'wix.iptf.demo.wishlist',
    'wix.links.v1.universal_link',
    'wix.live_site_data.v1.site_embed',
    'wix.locations.v1.location-schedules-location-events-dead-letters',
    'wix.loyalty_accounts.v1.LoyaltyAccount',
    'wix.members.activitycounters',
    'wix.members.follow',
    'wix.members.member',
    'wix.members.profiles.profile',
    'wix.membertags.api.v1.MemberTag',
    'wix.moderation.v2.moderated-content',
    'wix.moderation.v2.moderated-content.10657898-9841-4a92-ae07-9a3963c7617a',
    'wix.moderation.v2.moderated-content.10657898-9841-4a92-ae07-9a3963c7617a.wix.primeportal.post',
    'wix.moderation.v2.moderated-content.14bcded7-0066-7c35-14d7-466cb3f09103',
    'wix.moderation.v2.moderated-content.807bee44-f538-4440-be7f-be42309b5ef2.test.moderation.post',
    'wix.moderation.v2.moderated-content.8ac4ddf5-974a-4962-90f9-417d37cd0f48.test.local.moderation.post',
    'wix.moderation.v2.moderated-content.8c32640d-e73d-4a15-91e2-d9b7a89a98f6.wix.diners.v1.restaurant',
    'wix.multilingual.localization.localized_content',
    'wix.multilingual.v1.localization-public.localized_published_content',
    'wix.multilingual.v1.localization.localized_content',
    'wix.partners.loyalty.expert_account',
    'wix.partners.marketplace.contact',
    'wix.partners.marketplace.v1.provider-state',
    'wix.partners.partner_account',
    'wix.partners.partner_loyalty_activity',
    'wix.partners.settings.partner_settings',
    'wix.payments.v4.balances.balance',
    'wix.pbc.test',
    'wix.pos.v1.drawer.cash_report',
    'wix.pos.v1.staff_member.clock',
    'wix.pricing_plans.order',
    'wix.prime.auto.conversion_rule',
    'wix.prime.auto.editor_environment',
    'wix.prime.auto.example',
    'wix.prime.auto.membership',
    'wix.prime.auto.partner_connected_site',
    'wix.prime.auto.spider',
    'wix.prime.auto.wishlist',
    'wix.prime.auto.wishlistentry',
    'wix.pro_gallery.gallery',
    'wix.pro_gallery.gallery_v2',
    'wix.promote.fbe_mi.v1.integration',
    'wix.promote.marketing.ads_txt',
    'wix.promote.seo.robotsTxt',
    'wix.promote.seo.robots_txt',
    'wix.promote.url-mapper.urlMapping',
    'wix.promote.url-mapper.urlMappings',
    'wix.promote_campaigns_manager.v2.account',
    'wix.promote_campaigns_manager.v2.facebook_account',
    'wix.reports.report',
    'wix.restaurants.orders',
    'wix.restaurants.platform.api.order',
    'wix.restaurants.v2_1.contacts.contact',
    'wix.schema.protobuf_file_set',
    'wix.serverless.deployer.Artifact',
    'wix.sharedgallery.api.v1.items',
    'wix.sm.v1.member.status_changed',
    'wix.social.groups.group',
    'wix.social.groups.group_member',
    'wix.social.groups.group_request',
    'wix.social.groups.group_role',
    'wix.social.groups.groups_app_settings',
    'wix.social.groups.join_group_request',
    'wix.spamfilter.v1.spam_filter',
    'wix.stores.channels.taxonomy.v1.record',
    'wix.stores.channels.v1.facebook.token',
    'wix.stores.channels.wish.v1.token',
    'wix.stores.orders.order',
    'wix.url-mapper.urlMapping',
    'wix.url-mapper.urlMapping-wixStores-catalog-dead-letters',
    'wix.url-mapper.urlMapping-wixstores-catalog-wixsearch-indexer-dead-letters',
    'wix.virtual_numbers.blocked_numbers.v1.blocked_number',
    'wix.virtual_numbers.communication.records.call_record',
    'wix.virtual_numbers.switchboard.communication_record_changed',
    'wix.albumsproofing.CommentInFavouriteList',
    'wix.ecom.v1.shipping.rule',
    'wix.esb.configuration',
    'wix.the.button',
    'wix.crm.action_triggers.rule',
    'wix.crm.action_triggers.automation_template',
    'wix.crm.action_triggers.trigger_configuration',
    'wix.crm.action_triggers.automation_template_group',
    'wix.crm.action_triggers.state',
    'com.wixpress.wixdocs.project',
]);
exports.rule = {
    description: 'A message old_fqdn_for_backwards_compatibility annotation must be approved and whitelisted',
    severity: flynt_engine_1.Severity.Error,
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/fqdn#platformization-guidelines_fqdn_renaming-fqdns',
    run(ctx) {
        ctx.visit({
            visitMessage(message) {
                const entityOption = message.options.getTypedOption(wix_protos_1.wix.api.entity);
                const oldFqdn = entityOption && entityOption.value.oldFqdnForBackwardsCompatibility;
                if (oldFqdn) {
                    ctx.assert(whitelistOfLegacyFqdns.has(oldFqdn) || !(0, invalid_message_fqdn_name_1.isValidFqdn)(oldFqdn), {
                        errorAt: entityOption,
                        message: `the option 'old_fqdn_for_backwards_compatibility' with value '${oldFqdn}' in the message '${(0, fqdn_utils_1.toLocalFqn)(message.fqn)}' must be approved and whitelisted`
                    });
                }
            }
        });
    }
};
