// @wix/flynt-rules/dist/catalog/spi/spi-host-config/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'An SPI service must have a valid config',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2022-01-24'),
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if ((0, fqdn_utils_1.isSpiService)(service)) {
                    const spiConfigOptions = service.options.getTypedOption(wix_protos_1.wix.spi.service);
                    if (spiConfigOptions) {
                        const messageFqn = spiConfigOptions.value.configMessage;
                        if (messageFqn) {
                            ctx.assertTypeExists(messageFqn, flynt_engine_1.TypeName.Message, {
                                errorAt: spiConfigOptions,
                                message: 'SPI config-message must be an existing message'
                            });
                        }
                        else {
                            ctx.assert(messageFqn, {
                                errorAt: spiConfigOptions,
                                message: 'config_message attribute of wix.spi.service must be set'
                            });
                        }
                    }
                }
            }
        });
    }
};
