// @wix/flynt-plugins-host/index.ts
import {CliOption, FlyntOptions, FlyntOutput, FlyntPlugin, FlyntResults, FormatSeverity} from '@wix/flynt-plugin-types';
import {FlyntPluginsHostImpl} from './lib/host';
import {FlyntPluginsCliHostImpl} from './lib/cli-host';

export interface FlyntHostCtx {
  opts: FlyntOptions;
  debug?(text: string): void;
  reportError?(error: unknown): Promise<void>;
}

export interface PluginsHost {
  loadPlugin(plugin: FlyntPlugin): Promise<void>;
  createPluginsRunner(ctx: FlyntHostCtx): FlyntPlugins;
}

export interface PluginsCliHost extends PluginsHost {
  loadPluginsFromCli(args: string[]): Promise<void>;
  getExtendedCliOptions(): Record<string, CliOption[]>;
  setPluginOptions(plugin: string, opts: Record<string, any>): void;
}

export interface FlyntPlugins {
  init(): Promise<void>;
  beforeUnload(results: FlyntResults): Promise<void>;
  beforeValidateFlyntOptions(): FlyntOptions;
  beforeWriteToOutputFile(output: FlyntOutput): Promise<void>;
  afterWriteToOutputFile(output: FlyntOutput): Promise<void>;
  getResultsBaseFolder(): string | undefined;
  getFormatSeverityFunction(): FormatSeverity | undefined;
}

export function createFlyntPluginsHost(): PluginsHost {
  return new FlyntPluginsHostImpl();
}

export function createFlyntPluginsCliHost(): PluginsCliHost {
  return new FlyntPluginsCliHostImpl();
}
