// @wix/flynt-rules/dist/catalog/string-format-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSnakeCase = exports.isFirstUpperCase = exports.isUpperCamelCase = exports.isAllLowerCase = exports.isLowerSnakeCase = exports.isKebabCase = exports.isAllUpperCase = exports.isAllUpperCaseSnakeCase = void 0;
function isAllUpperCaseSnakeCase(word) {
    return isAllUpperCase(word) && !isKebabCase(word);
}
exports.isAllUpperCaseSnakeCase = isAllUpperCaseSnakeCase;
function isAllUpperCase(word) {
    return word === word.toUpperCase();
}
exports.isAllUpperCase = isAllUpperCase;
function isKebabCase(word) {
    return /[-]/.test(word);
}
exports.isKebabCase = isKebabCase;
function isLowerSnakeCase(word) {
    return isAllLowerCase(word) && !isKebabCase(word);
}
exports.isLowerSnakeCase = isLowerSnakeCase;
function isAllLowerCase(word) {
    return word === word.toLowerCase();
}
exports.isAllLowerCase = isAllLowerCase;
function isUpperCamelCase(word) {
    return isFirstUpperCase(word) && !isAllUpperCase(word) && !isKebabCase(word) && !isSnakeCase(word);
}
exports.isUpperCamelCase = isUpperCamelCase;
function isFirstUpperCase(word) {
    return /[A-Z]/.test(word.charAt(0));
}
exports.isFirstUpperCase = isFirstUpperCase;
function isSnakeCase(word) {
    return /[_]/.test(word);
}
exports.isSnakeCase = isSnakeCase;
