// @wix/proto-syntax-tree/dist/lib/visitor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.visitorContainerMethods = exports.visitorMethods = exports.skip = exports.visitSyntaxTreeFrom = exports.visitSyntaxTree = void 0;
var iterator_1 = require("./iterator");
Object.defineProperty(exports, "visitSyntaxTree", { enumerable: true, get: function () { return iterator_1.visitSyntaxTree; } });
Object.defineProperty(exports, "visitSyntaxTreeFrom", { enumerable: true, get: function () { return iterator_1.visitSyntaxTreeFrom; } });
var skipper_1 = require("./skipper");
Object.defineProperty(exports, "skip", { enumerable: true, get: function () { return skipper_1.skip; } });
const vm = {
    visitFile: true,
    visitService: true,
    visitMessage: true,
    visitEnum: true,
    visitMethod: true,
    visitField: true,
    visitOneof: true,
    visitEnumValue: true,
    visitExtension: true,
};
const vcm = {
    visitFile: true,
    visitService: true,
    visitMessage: true,
    visitEnum: true,
};
exports.visitorMethods = Object.keys(vm);
exports.visitorContainerMethods = Object.keys(vcm);
