// @wix/flynt-rules/dist/catalog/fqdns/old-fqdn-for-backward-compatibility-rename-not-allowed/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'old_fqdn_for_backwards_compatibility is immutable',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/fqdn#platformization-guidelines_fqdn_hard-path',
    stateful: true,
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visitModified({
            visitMessage(message, prevMessage) {
                const prevBackwardCompatibilityFqdn = prevMessage.options.getTypedOption(wix_protos_1.wix.api.entity)?.value.oldFqdnForBackwardsCompatibility;
                if (prevBackwardCompatibilityFqdn) {
                    const currentBackwardCompatibilityFqdn = message.options.getTypedOption(wix_protos_1.wix.api.entity)?.value.oldFqdnForBackwardsCompatibility;
                    ctx.assert(currentBackwardCompatibilityFqdn === prevBackwardCompatibilityFqdn, {
                        errorAt: message,
                        message: 'changing of the "old_fqdn_for_backwards_compatibility" value is not allowed'
                    });
                }
            }
        });
    }
};
