// @wix/flynt-rules/dist/catalog/fqdns/missing-service-entity/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'A service must have a service entity',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/fqdn#platformization-guidelines_fqdn_annotate-the-service-as-related-to-that-entity',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if ((0, fqdn_utils_1.isPrivateExposure)(service) || service.methods.length === 0) {
                    return;
                }
                const spiOption = service.options.getTypedOption(wix_protos_1.wix.spi.service);
                if (spiOption && spiOption.value.fqdn) {
                    return;
                }
                ctx.assert((0, fqdn_utils_1.resolveMainEntityFqn)(service), {
                    errorAt: service,
                    message: `the service '${service.name}' must have the 'wix.api.service_entity' option with fqn of the service entity`
                });
            }
        });
    }
};
