// @wix/flynt-rules/dist/catalog/fqdns/invalid-message-fqdn-name/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = exports.isValidFqdn = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
const FQDN_REGEX = /^wix\.[a-z]+(_[a-z]+)*(\.[a-z]+(_[a-z]+)*)?\.v\d{1,3}\.[a-z]+(_[a-z]+)*$/;
const FQDN_MAX_LENGTH = 128;
function isValidFqdn(fqdn) {
    return FQDN_REGEX.test(fqdn);
}
exports.isValidFqdn = isValidFqdn;
exports.rule = {
    description: 'A message fqdn annotation must be in a valid format',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/fqdn#platformization-guidelines_fqdn_fqdn-pattern',
    severity: flynt_engine_1.Severity.Error,
    graceUntil: new Date('2021-12-20'),
    run(ctx) {
        ctx.visit({
            visitMessage(message) {
                const entityOption = message.options.getTypedOption(wix_protos_1.wix.api.entity);
                const fqdn = entityOption && entityOption.value.fqdn;
                if (fqdn) {
                    ctx.assert(isValidFqdn(fqdn), {
                        errorAt: entityOption,
                        message: `the message '${(0, fqdn_utils_1.toLocalFqn)(message.fqn)}' has an invalid fqdn '${fqdn}'`
                    });
                    ctx.assert(fqdn.length <= FQDN_MAX_LENGTH, {
                        errorAt: entityOption,
                        message: `the message '${(0, fqdn_utils_1.toLocalFqn)(message.fqn)}' has an invalid fqdn length of ${fqdn.length} chars`
                    });
                }
            }
        });
    }
};
