// @wix/flynt-rules/dist/catalog/spi/spi-post-methods-only/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const flynt_engine_1 = require("@wix/flynt-engine");
const fqdn_utils_1 = require("../../fqdn-utils");
const wix_protos_1 = require("@wix/wix-protos");
exports.rule = {
    description: 'Only POST methods are allowed in SPIs',
    moreInfoUrl: 'https://bo.wix.com/wix-docs/rnd/platformization-guidelines/spi---service-provider-interface#platformization-guidelines_spi---service-provider-interface_googleapihttp',
    severity: flynt_engine_1.Severity.Error,
    run(ctx) {
        ctx.visit({
            visitService(service) {
                if (!(0, fqdn_utils_1.isSpiService)(service)) {
                    return false;
                }
            },
            visitMethod(method) {
                const httpOption = method.options.getTypedOption(wix_protos_1.google.api.http);
                ctx.assert(httpOption && httpOption.value.post, {
                    errorAt: method,
                    message: 'Only POST methods are allowed in SPIs'
                });
            }
        });
    }
};
