// @wix/wix-protos/typed/wix/api/api_errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusCodes = void 0;
var StatusCodes;
(function (StatusCodes) {
    let HttpStatusCode;
    (function (HttpStatusCode) {
        HttpStatusCode[HttpStatusCode["OK"] = 0] = "OK";
        HttpStatusCode[HttpStatusCode["UNAUTHENTICATED"] = 1] = "UNAUTHENTICATED";
        HttpStatusCode[HttpStatusCode["PERMISSION_DENIED"] = 2] = "PERMISSION_DENIED";
        HttpStatusCode[HttpStatusCode["INVALID_ARGUMENT"] = 3] = "INVALID_ARGUMENT";
        HttpStatusCode[HttpStatusCode["NOT_FOUND"] = 4] = "NOT_FOUND";
        HttpStatusCode[HttpStatusCode["INTERNAL"] = 5] = "INTERNAL";
        HttpStatusCode[HttpStatusCode["UNAVAILABLE"] = 6] = "UNAVAILABLE";
        HttpStatusCode[HttpStatusCode["RESOURCE_EXHAUSTED"] = 7] = "RESOURCE_EXHAUSTED";
        HttpStatusCode[HttpStatusCode["CANCELED"] = 8] = "CANCELED";
        HttpStatusCode[HttpStatusCode["ALREADY_EXISTS"] = 9] = "ALREADY_EXISTS";
        HttpStatusCode[HttpStatusCode["FAILED_PRECONDITION"] = 10] = "FAILED_PRECONDITION";
    })(HttpStatusCode = StatusCodes.HttpStatusCode || (StatusCodes.HttpStatusCode = {}));
})(StatusCodes = exports.StatusCodes || (exports.StatusCodes = {}));
