// @wix/wix-protos/typed/wix/api/callback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Segment = exports.Callback = void 0;
var Callback;
(function (Callback) {
    let EventType;
    (function (EventType) {
        EventType[EventType["UNKNOWN"] = 0] = "UNKNOWN";
        EventType[EventType["CREATED"] = 1] = "CREATED";
        EventType[EventType["UPDATED"] = 2] = "UPDATED";
        EventType[EventType["DELETED"] = 3] = "DELETED";
        EventType[EventType["ACTION"] = 4] = "ACTION";
    })(EventType = Callback.EventType || (Callback.EventType = {}));
})(Callback = exports.Callback || (exports.Callback = {}));
var Segment;
(function (Segment_1) {
    let Segment;
    (function (Segment) {
        Segment[Segment["UNSPECIFIED"] = 0] = "UNSPECIFIED";
        Segment[Segment["USERS"] = 1] = "USERS";
        Segment[Segment["PUBLIC"] = 2] = "PUBLIC";
        Segment[Segment["OTHERS"] = 3] = "OTHERS";
    })(Segment = Segment_1.Segment || (Segment_1.Segment = {}));
})(Segment = exports.Segment || (exports.Segment = {}));
