// @wix/wix-protos/typed/wix/api/validations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Format = void 0;
var Format;
(function (Format) {
    Format[Format["EMAIL"] = 0] = "EMAIL";
    Format[Format["HOSTNAME"] = 1] = "HOSTNAME";
    Format[Format["IPV4"] = 2] = "IPV4";
    Format[Format["IPV6"] = 3] = "IPV6";
    Format[Format["URI"] = 4] = "URI";
    Format[Format["WEB_URL"] = 5] = "WEB_URL";
    Format[Format["PHONE"] = 7] = "PHONE";
    Format[Format["CREDIT_CARD"] = 8] = "CREDIT_CARD";
    Format[Format["GUID"] = 9] = "GUID";
    Format[Format["COUNTRY"] = 10] = "COUNTRY";
    Format[Format["LANGUAGE"] = 11] = "LANGUAGE";
    Format[Format["DECIMAL_VALUE"] = 12] = "DECIMAL_VALUE";
    Format[Format["CURRENCY"] = 13] = "CURRENCY";
    Format[Format["LANGUAGE_TAG"] = 14] = "LANGUAGE_TAG";
    Format[Format["COLOR_HEX"] = 15] = "COLOR_HEX";
    Format[Format["LOCAL_DATE"] = 16] = "LOCAL_DATE";
    Format[Format["LOCAL_TIME"] = 17] = "LOCAL_TIME";
    Format[Format["LOCAL_DATE_TIME"] = 18] = "LOCAL_DATE_TIME";
})(Format = exports.Format || (exports.Format = {}));
