// @wix/wnp-proto-validation-fv/lib/fv/messages.js
module.exports = {
  required: '\'{field}\' is required',

  string: '\'{field}\' must be a string',
  stringEmpty: '\'{field}\' must not be empty',
  stringMin: '\'{field}\' length must be greater than or equal to {expected} characters long',
  stringMax: '\'{field}\' length must be less than or equal to {expected} characters long',
  stringLength: '\'{field}\' length must be {expected} characters long',
  stringPattern: '\'{field}\' fails to match the required pattern',
  stringContains: '\'{field}\' must contain the \'{expected}\' text',
  stringEnum: '\'{field}\' does not match any of the allowed values',
  stringNumeric: '\'{field}\' must be a numeric string',
  stringAlpha: '\'{field}\' must be an alphabetic string',
  stringAlphanum: '\'{field}\' must be an alphanumeric string',
  stringAlphadash: '\'{field}\' must be an alphadash string',
  stringHex: '\'{field}\' must be a hex string',
  stringSingleLine: '\'{field}\' must be a single line string',
  stringBase64: '\'{field}\' must be a base64 string',

  number: '\'{field}\' must be a number',
  numberMin: '\'{field}\' must be greater than or equal to {expected}',
  numberMax: '\'{field}\' must be less than or equal to {expected}',
  numberEqual: '\'{field}\' must be equal to {expected}',
  numberNotEqual: '\'{field}\' can\'t be equal to {expected}',
  numberInteger: '\'{field}\' must be an integer',
  numberPositive: '\'{field}\' must be a positive number',
  numberNegative: '\'{field}\' must be a negative number',

  array: '\'{field}\' must be an array',
  arrayEmpty: '\'{field}\' must not be an empty array',
  arrayMin: '\'{field}\' must contain at least {expected} items',
  arrayMax: '\'{field}\' must contain less than or equal to {expected} items',
  arrayLength: '\'{field}\' must contain {expected} items',
  arrayContains: '\'{field}\' must contain the \'{expected}\' item',
  arrayUnique: '\'{actual}\' value in \'{field}\' does not unique the \'{expected}\' values',
  arrayEnum: '\'{actual}\' value in \'{field}\' does not match any of the \'{expected}\' values',

  tuple: '\'{field}\' must be an array',
  tupleEmpty: '\'{field}\' must not be an empty array',
  tupleLength: '\'{field}\' must contain {expected} items',

  boolean: '\'{field}\' must be a boolean',

  currency: '\'{field}\' must be a valid currency format',

  date: '\'{field}\' must be a Date',
  dateMin: '\'{field}\' must be greater than or equal to {expected}',
  dateMax: '\'{field}\' must be less than or equal to {expected}',

  enumValue: '\'{field}\' value \'{expected}\' does not match any of the allowed values',

  equalValue: '\'{field}\' value must be equal to \'{expected}\'',
  equalField: '\'{field}\' value must be equal to \'{expected}\' value',

  forbidden: '\'{field}\' is forbidden',

  function: '\'{field}\' must be a function',

  email: '\'{field}\' must be a valid e-mail',
  emailEmpty: '\'{field}\' must not be empty',
  emailMin: '\'{field}\' length must be greater than or equal to {expected} characters long',
  emailMax: '\'{field}\' length must be less than or equal to {expected} characters long',

  luhn: '\'{field}\' must be a valid credit card number',

  mac: '\'{field}\' must be a valid MAC address',

  object: '\'{field}\' must be an object',
  objectStrict: 'object \'{field}\' contains forbidden keys: \'{actual}\'',
  objectMinProps: 'object \'{field}\' must contain at least {expected} properties',
  objectMaxProps: 'object \'{field}\' must contain {expected} properties at most',

  url: '\'{field}\' must be a valid URL',
  urlEmpty: '\'{field}\' must not be empty',

  uuid: '\'{field}\' must be a valid UUID',
  uuidVersion: '\'{field}\' must be a valid UUID version provided',

  classInstanceOf: '\'{field}\' must be an instance of the \'{expected}\' class',

  objectID: '\'{field}\' must be an valid ObjectID',
};
