// @wix/proto-syntax-tree/dist/lib/syntax-tree/descriptor-paths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneofDescriptorPath = exports.FieldDescriptorPath = exports.EnumValueDescriptorPath = exports.EnumDescriptorPath = exports.MessageDescriptorPath = exports.MethodDescriptorPath = exports.ServiceDescriptorPath = exports.FileDescriptorPath = void 0;
var FileDescriptorPath;
(function (FileDescriptorPath) {
    FileDescriptorPath[FileDescriptorPath["package"] = 2] = "package";
    FileDescriptorPath[FileDescriptorPath["message"] = 4] = "message";
    FileDescriptorPath[FileDescriptorPath["enum"] = 5] = "enum";
    FileDescriptorPath[FileDescriptorPath["service"] = 6] = "service";
    FileDescriptorPath[FileDescriptorPath["extensions"] = 7] = "extensions";
    FileDescriptorPath[FileDescriptorPath["options"] = 8] = "options";
    FileDescriptorPath[FileDescriptorPath["syntax"] = 12] = "syntax";
})(FileDescriptorPath = exports.FileDescriptorPath || (exports.FileDescriptorPath = {}));
var ServiceDescriptorPath;
(function (ServiceDescriptorPath) {
    ServiceDescriptorPath[ServiceDescriptorPath["name"] = 1] = "name";
    ServiceDescriptorPath[ServiceDescriptorPath["method"] = 2] = "method";
    ServiceDescriptorPath[ServiceDescriptorPath["options"] = 3] = "options";
})(ServiceDescriptorPath = exports.ServiceDescriptorPath || (exports.ServiceDescriptorPath = {}));
var MethodDescriptorPath;
(function (MethodDescriptorPath) {
    MethodDescriptorPath[MethodDescriptorPath["name"] = 1] = "name";
    MethodDescriptorPath[MethodDescriptorPath["input"] = 2] = "input";
    MethodDescriptorPath[MethodDescriptorPath["output"] = 3] = "output";
    MethodDescriptorPath[MethodDescriptorPath["options"] = 4] = "options";
})(MethodDescriptorPath = exports.MethodDescriptorPath || (exports.MethodDescriptorPath = {}));
var MessageDescriptorPath;
(function (MessageDescriptorPath) {
    MessageDescriptorPath[MessageDescriptorPath["name"] = 1] = "name";
    MessageDescriptorPath[MessageDescriptorPath["field"] = 2] = "field";
    MessageDescriptorPath[MessageDescriptorPath["message"] = 3] = "message";
    MessageDescriptorPath[MessageDescriptorPath["enum"] = 4] = "enum";
    MessageDescriptorPath[MessageDescriptorPath["options"] = 7] = "options";
    MessageDescriptorPath[MessageDescriptorPath["oneof"] = 8] = "oneof";
})(MessageDescriptorPath = exports.MessageDescriptorPath || (exports.MessageDescriptorPath = {}));
var EnumDescriptorPath;
(function (EnumDescriptorPath) {
    EnumDescriptorPath[EnumDescriptorPath["name"] = 1] = "name";
    EnumDescriptorPath[EnumDescriptorPath["value"] = 2] = "value";
    EnumDescriptorPath[EnumDescriptorPath["options"] = 3] = "options";
})(EnumDescriptorPath = exports.EnumDescriptorPath || (exports.EnumDescriptorPath = {}));
var EnumValueDescriptorPath;
(function (EnumValueDescriptorPath) {
    EnumValueDescriptorPath[EnumValueDescriptorPath["name"] = 1] = "name";
    EnumValueDescriptorPath[EnumValueDescriptorPath["number"] = 2] = "number";
    EnumValueDescriptorPath[EnumValueDescriptorPath["options"] = 3] = "options";
})(EnumValueDescriptorPath = exports.EnumValueDescriptorPath || (exports.EnumValueDescriptorPath = {}));
var FieldDescriptorPath;
(function (FieldDescriptorPath) {
    FieldDescriptorPath[FieldDescriptorPath["name"] = 1] = "name";
    FieldDescriptorPath[FieldDescriptorPath["number"] = 3] = "number";
    FieldDescriptorPath[FieldDescriptorPath["label"] = 4] = "label";
    FieldDescriptorPath[FieldDescriptorPath["type"] = 5] = "type";
    FieldDescriptorPath[FieldDescriptorPath["typeName"] = 6] = "typeName";
    FieldDescriptorPath[FieldDescriptorPath["options"] = 8] = "options";
})(FieldDescriptorPath = exports.FieldDescriptorPath || (exports.FieldDescriptorPath = {}));
var OneofDescriptorPath;
(function (OneofDescriptorPath) {
    OneofDescriptorPath[OneofDescriptorPath["name"] = 1] = "name";
    OneofDescriptorPath[OneofDescriptorPath["options"] = 2] = "options";
})(OneofDescriptorPath = exports.OneofDescriptorPath || (exports.OneofDescriptorPath = {}));
