// @wix/proto-syntax-tree/dist/lib/syntax-tree/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumImpl = void 0;
const api_1 = require("../api");
const base_type_1 = require("./base-type");
const enumValue_1 = require("./enumValue");
const descriptor_paths_1 = require("./descriptor-paths");
const options_1 = require("./options");
const source_code_index_1 = require("./source-code-index");
const source_code_details_1 = require("./source-code-details");
const reserved_1 = require("./reserved");
class EnumImpl extends base_type_1.BaseType {
    constructor(st, descriptor, parent, pathInfo, sourceIdx) {
        super(descriptor, parent);
        this.st = st;
        this.typeName = api_1.TypeName.Enum;
        this.descriptor = descriptor;
        this.parent = parent;
        this.reserved = new reserved_1.ReservedWithInclusiveRange(descriptor);
        this.options = new options_1.OptionsImpl(st, { options: descriptor.options, optionType: api_1.OptionType.EnumOptions }, this, (0, source_code_index_1.pInfo)(pathInfo, descriptor_paths_1.EnumDescriptorPath.options), sourceIdx);
        this.detailedSourceCode = (0, source_code_details_1.initDetailedSourceCode)(pathInfo, sourceIdx, this, {
            'name': descriptor_paths_1.EnumDescriptorPath.name,
        });
        st.addType(this);
        this.values = [];
        let i = 0;
        for (const enumValueDescriptor of descriptor.valueList) {
            const enumValue = new enumValue_1.EnumValueImpl(st, enumValueDescriptor, this, (0, source_code_index_1.pInfo)(pathInfo, descriptor_paths_1.EnumDescriptorPath.value, i++), sourceIdx);
            this.values.push(enumValue);
        }
    }
    get file() {
        return this.parent.file;
    }
    toJSON() {
        return {
            typeName: this.typeName,
            name: this.name,
            options: this.options.allOptions,
            values: this.values,
        };
    }
}
exports.EnumImpl = EnumImpl;
