// @wix/proto-syntax-tree/dist/lib/syntax-tree/enumValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumValueImpl = void 0;
const api_1 = require("../api");
const base_type_1 = require("./base-type");
const options_1 = require("./options");
const descriptor_paths_1 = require("./descriptor-paths");
const source_code_index_1 = require("./source-code-index");
const source_code_details_1 = require("./source-code-details");
class EnumValueImpl extends base_type_1.BaseType {
    constructor(st, descriptor, parent, pathInfo, sourceIdx) {
        super(descriptor, parent);
        this.st = st;
        this.typeName = api_1.TypeName.EnumValue;
        this.descriptor = descriptor;
        this.parent = parent;
        this.options = new options_1.OptionsImpl(st, { options: descriptor.options, optionType: api_1.OptionType.EnumValueOptions }, this, (0, source_code_index_1.pInfo)(pathInfo, descriptor_paths_1.EnumValueDescriptorPath.options), sourceIdx);
        this.detailedSourceCode = (0, source_code_details_1.initDetailedSourceCode)(pathInfo, sourceIdx, this, {
            'name': descriptor_paths_1.EnumValueDescriptorPath.name,
            'number': descriptor_paths_1.EnumValueDescriptorPath.number,
        });
        st.addType(this);
    }
    get file() {
        return this.parent.file;
    }
    get number() {
        return this.descriptor.number;
    }
    toJSON() {
        return {
            typeName: this.typeName,
            name: this.name,
            options: this.options.allOptions,
            number: this.number,
        };
    }
}
exports.EnumValueImpl = EnumValueImpl;
