// @wix/proto-syntax-tree/dist/lib/syntax-tree/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionImpl = void 0;
const base_type_1 = require("./base-type");
const field_type_1 = require("./field-type");
const api_1 = require("../api");
const source_code_details_1 = require("./source-code-details");
class ExtensionImpl extends base_type_1.BaseType {
    constructor(st, descriptor, parent, pathInfo, sourceIdx) {
        super(descriptor, parent);
        this.st = st;
        this.typeName = api_1.TypeName.Extension;
        this.descriptor = descriptor;
        this.parent = parent;
        this.optionType = toOptionType(descriptor.extendee);
        this.detailedSourceCode = (0, source_code_details_1.initDetailedSourceCode)(pathInfo, sourceIdx, this);
        st.addType(this);
        st.addOption(this);
    }
    get fieldType() {
        if (!this._fieldType) {
            this._fieldType = field_type_1.FieldTypeFactory.createFieldTypeFromDescriptor(this.st, this.descriptor);
        }
        return this._fieldType;
    }
    get file() {
        return this.parent;
    }
    get number() {
        return this.descriptor.number;
    }
    toJSON() {
        return {
            typeName: this.typeName,
            name: this.name,
            number: this.number,
            fieldType: this.fieldType,
            optionType: this.optionType ? api_1.OptionType[this.optionType] : undefined,
        };
    }
}
exports.ExtensionImpl = ExtensionImpl;
function toOptionType(extendee) {
    switch (extendee) {
        case '.google.protobuf.FileOptions':
            return api_1.OptionType.FileOptions;
        case '.google.protobuf.ServiceOptions':
            return api_1.OptionType.ServiceOptions;
        case '.google.protobuf.MethodOptions':
            return api_1.OptionType.MethodOptions;
        case '.google.protobuf.MessageOptions':
            return api_1.OptionType.MessageOptions;
        case '.google.protobuf.EnumOptions':
            return api_1.OptionType.EnumOptions;
        case '.google.protobuf.FieldOptions':
            return api_1.OptionType.FieldOptions;
        case '.google.protobuf.OneofOptions':
            return api_1.OptionType.OneofOptions;
        case '.google.protobuf.EnumValueOptions':
            return api_1.OptionType.EnumValueOptions;
    }
    return undefined;
}
