// @wix/proto-syntax-tree/dist/lib/syntax-tree/one-of.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneofImpl = void 0;
const api_1 = require("../api");
const base_type_1 = require("./base-type");
const options_1 = require("./options");
const descriptor_paths_1 = require("./descriptor-paths");
const source_code_index_1 = require("./source-code-index");
const source_code_details_1 = require("./source-code-details");
class OneofImpl extends base_type_1.BaseType {
    constructor(st, descriptor, parent, pathInfo, sourceIdx) {
        super(descriptor, parent);
        this.st = st;
        this.typeName = api_1.TypeName.Oneof;
        this.fields = [];
        this.descriptor = descriptor;
        this.parent = parent;
        this.options = new options_1.OptionsImpl(st, { options: descriptor.options, optionType: api_1.OptionType.OneofOptions }, this, (0, source_code_index_1.pInfo)(pathInfo, descriptor_paths_1.OneofDescriptorPath.options), sourceIdx);
        this.detailedSourceCode = (0, source_code_details_1.initDetailedSourceCode)(pathInfo, sourceIdx, this, {
            'name': descriptor_paths_1.OneofDescriptorPath.name,
        });
        st.addType(this);
    }
    addField(field) {
        this.fields.push(field);
    }
    get file() {
        return this.parent.file;
    }
    toJSON() {
        return {
            typeName: this.typeName,
            name: this.name,
            options: this.options.allOptions,
            fields: this.fields.map(f => f.name),
        };
    }
}
exports.OneofImpl = OneofImpl;
