// @wix/proto-syntax-tree/dist/lib/syntax-tree/reserved.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReservedWithInclusiveRange = exports.ReservedWithExclusiveRange = void 0;
class BaseReservedImpl {
    constructor(descriptor) {
        this.descriptor = descriptor;
    }
    get reservedNames() {
        return this.descriptor.reservedNameList;
    }
    get reservedNumbers() {
        return this.descriptor.reservedRangeList;
    }
    isReservedName(name) {
        for (const resName of this.reservedNames) {
            if (resName === name) {
                return true;
            }
        }
        return false;
    }
    isReservedNumber(num) {
        for (const range of this.reservedNumbers) {
            if (this.isNumInRange(num, range)) {
                return true;
            }
        }
        return false;
    }
}
class ReservedWithExclusiveRange extends BaseReservedImpl {
    isNumInRange(num, range) {
        return num >= range.start && num < range.end;
    }
}
exports.ReservedWithExclusiveRange = ReservedWithExclusiveRange;
class ReservedWithInclusiveRange extends BaseReservedImpl {
    isNumInRange(num, range) {
        return num >= range.start && num <= range.end;
    }
}
exports.ReservedWithInclusiveRange = ReservedWithInclusiveRange;
