// @wix/proto-syntax-tree/dist/lib/syntax-tree/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceImpl = void 0;
const api_1 = require("../api");
const base_type_1 = require("./base-type");
const method_1 = require("./method");
const descriptor_paths_1 = require("./descriptor-paths");
const options_1 = require("./options");
const source_code_index_1 = require("./source-code-index");
const source_code_details_1 = require("./source-code-details");
class ServiceImpl extends base_type_1.BaseType {
    constructor(st, descriptor, parent, pathInfo, sourceIdx) {
        super(descriptor, parent);
        this.st = st;
        this.typeName = api_1.TypeName.Service;
        this.descriptor = descriptor;
        this.parent = parent;
        this.options = new options_1.OptionsImpl(st, { options: descriptor.options, optionType: api_1.OptionType.ServiceOptions }, this, (0, source_code_index_1.pInfo)(pathInfo, descriptor_paths_1.ServiceDescriptorPath.options), sourceIdx);
        this.detailedSourceCode = (0, source_code_details_1.initDetailedSourceCode)(pathInfo, sourceIdx, this, {
            'name': descriptor_paths_1.ServiceDescriptorPath.name,
        });
        st.addType(this);
        this.methods = [];
        let i = 0;
        for (const methodDescriptor of descriptor.methodList) {
            const method = new method_1.MethodImpl(st, methodDescriptor, this, (0, source_code_index_1.pInfo)(pathInfo, descriptor_paths_1.ServiceDescriptorPath.method, i++), sourceIdx);
            this.methods.push(method);
        }
    }
    get file() {
        return this.parent.file;
    }
    toJSON() {
        return {
            typeName: this.typeName,
            name: this.name,
            options: this.options.allOptions,
            methods: this.methods,
        };
    }
}
exports.ServiceImpl = ServiceImpl;
