// @wix/proto-syntax-tree/dist/lib/syntax-tree/source-code-details.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initDetailedSourceCode = void 0;
class DetailedSourceCodeImpl {
    constructor(pathInfo, sourceIdx, container, details) {
        this.pathInfo = pathInfo;
        this.sourceIdx = sourceIdx;
        this.container = container;
        if (details) {
            for (const [key, val] of Object.entries(details)) {
                if (Array.isArray(val)) {
                    Object.defineProperty(this, key, {
                        get: () => {
                            return this.sourceIdx.getSourceCodeInfoBySubPaths(this.pathInfo, this.container, key, val);
                        }
                    });
                }
                else {
                    Object.defineProperty(this, key, {
                        get: () => {
                            return this.sourceIdx.getSourceCodeInfoBySubPath(this.pathInfo, this.container, key, val);
                        }
                    });
                }
            }
        }
    }
    get mySourceCodeInfo() {
        return this.sourceIdx.getSourceCodeInfoByPath(this.pathInfo, this.container);
    }
}
function initDetailedSourceCode(pathInfo, sourceIdx, container, details) {
    return new DetailedSourceCodeImpl(pathInfo, sourceIdx, container, details);
}
exports.initDetailedSourceCode = initDetailedSourceCode;
